clientcontroller/opstackl2/consumer.go

```diff
@@ -431,6 +433,48 @@ func ConvertProof(cmtProof cmtcrypto.Proof) Proof {
 	}
 }
 
+// GetBlockNumberByTimestamp returns the L2 block number for the given BTC staking activation timestamp.
+// It uses a binary search to find the block number.
+func (cc *OPStackL2ConsumerController) GetBlockNumberByTimestamp(ctx context.Context, targetTimestamp uint64) (uint64, error) {
+	// Check if the target timestamp is after the latest block
+	latestBlock, err := cc.opl2Client.HeaderByNumber(ctx, nil)
+	if err != nil {
+		return 0, err
+	}
+	if targetTimestamp > latestBlock.Time {
+		return 0, nil
+	}
+	// Check if the target timestamp is before the first block
+	firstBlock, err := cc.opl2Client.HeaderByNumber(ctx, big.NewInt(1))
+	if err != nil {
+		return 0, err
+	}
+	if targetTimestamp < firstBlock.Time {
+		return 0, nil
+	}
+
+	lowerBound := uint64(1)
```

Comment:
i am also curious, why not compare with block 0 directly but pick block 1?

Reply:
b/c calling the `HeaderByNumber` of block 0, its timestamp is 0

```
curl -X POST https://rpc.ankr.com/optimism -d '{"jsonrpc":"2.0","method":"eth_getBlockByNumber","params":["0x0", false],"id":1}'

{"jsonrpc":"2.0","id":1,"result":{"hash":"0x7ca38a1916c42007829c55e69d3e9a73265554b586a499015373241b8a3fa48b","number":"0x0","timestamp":"0x0","transactions":[]}}
```